kong/tools/public.lua

```diff
@@ -12,62 +12,139 @@ do
+  }
+
+
+  _M.req_mime_types  = MIME_TYPES
+  _M.req_body_errors = ERRORS
+
+
+  local MIME_DECODERS = {
+    [MIME_TYPES.multipart] = function(content_type)
+      local raw_body = ngx_req_get_body_data()
+      local args     = multipart(raw_body, content_type):get_all()
+
+      return args, raw_body
     end,
 
-    [2] = function() -- json
-      local body, err = cjson.decode(ngx_req_get_body_data())
+    [MIME_TYPES.json] = function()
+      local raw_body  = ngx_req_get_body_data()
+      local args, err = cjson.decode(raw_body)
       if err then
         ngx_log(ERR, "could not decode JSON body args: ", err)
-        return {}
+        return {}, raw_body
       end
 
-      return body
+      return args, raw_body
     end,
 
-    [3] = function() -- encoded form
+    [MIME_TYPES.form_url_encoded] = function()
       local ok, res, err = pcall(ngx_req_get_post_args)
       if not ok or err then
         local msg = res and res or err
         ngx_log(ERR, "could not get body args: ", msg)
         return {}
       end
 
+      -- don't read raw_body if not necessary
+      -- if we called get_body_args(), we only want the parsed body
       return res
     end,
   }
 
-  function _M.get_body_args()
+
+  local function get_body_info()
     local content_type = ngx.var.http_content_type
 
     if not content_type or content_type == "" then
-      return {}
+      ngx_log(ERR, ERRORS[ERRORS.no_ct])
+
+      return {}, ERRORS.no_ct
     end
 
-    local map_type
+    local req_mime
 
     if str_find(content_type, "multipart/form-data", nil, true) then
-      map_type = 1
+      req_mime = MIME_TYPES.multipart
 
     elseif str_find(content_type, "application/json", nil, true) then
-      map_type = 2
+      req_mime = MIME_TYPES.json
 
     elseif str_find(content_type, "application/www-form-urlencoded", nil, true) or
       str_find(content_type, "application/x-www-form-urlencoded", nil, true) then
-      map_type = 3
+      req_mime = MIME_TYPES.form_url_encoded
+
+    elseif str_find(content_type, "application/xml", nil, true) then
+      req_mime = MIME_TYPES.xml
+
+    elseif str_find(content_type, "text/plain", nil, true) then
+      req_mime = MIME_TYPES.text
+
+    elseif str_find(content_type, "text/html", nil, true) then
+      req_mime = MIME_TYPES.html
+    end
+
+    if not req_mime then
+      -- unknown Content-Type
+      ngx_log(ERR, str_format(ERRORS[ERRORS.unsupported_ct], content_type))
+
+      return {}, ERRORS.unknown_ct
+    end
+
+    if not MIME_DECODERS[req_mime] then
+      -- known Content-Type, but cannot decode
+      ngx_log(ERR, str_format(ERRORS[ERRORS.unsupported_ct], content_type))
+
+      return {}, ERRORS.unsupported_ct, nil, req_mime
+    end
+
+    -- decoded Content-Type
+    local args, raw_body = MIME_DECODERS[req_mime](content_type)
+
+    return args, nil, raw_body, req_mime
+  end
+
+
+  function _M.get_body_args()
+    -- args
+    return (get_body_info())
```

Comment:
Wrapping this in parens removes the tail call, iirc. Assuming this is intentional behavior?